### PR TITLE
Ensure that we bind collection initializer elements, even in presence…

### LINKED ISSUE #18074: Collection initializer in presence of semantic errors doesn't contain operations for child nodes
Consider the below source code:

```
Imports System.Collections.Generic
Imports System.Linq
Imports System.Threading.Tasks

Class Program
    Private Shared Sub Main(args As String())
        Dim x = New Integer(1, 1) {{{1, 1}}, {2, 2}}'BIND:"{1, 1}"
    End Sub
End Class
```

I would expect the IOperation treefor `{1, 1}` to return IOperation nodes for bound literals 1 and 1 within the collection initializer `{1, 1}`. This doesn't seem to be the case:

```
IOperation tree for "{{1, 1}}"

IArrayInitializer (1 elements) (OperationKind.ArrayInitializer, IsInvalid)
  IInvalidExpression (OperationKind.InvalidExpression, Type: ?, IsInvalid)
```

Now try to get IOperation for one of it's child node:
```
IOperation tree for "1"

ILiteralExpression (Text: 1) (OperationKind.LiteralExpression, Type: System.Int32, Constant: 1)
```

### PR DESCRIPTION
**Pull request by mavasani**
… of errors

This change fixes the VB binder, C# already does it correctly.
Added unit tests for both VB and C#.
Fixes #18074

### CODE REVIEW

**Comment by 333fred** [comment]
@mavasani test failures look legitimate

File: src/Compilers/VisualBasic/Portable/Binding/Binder_Expressions.vb [hunk 1]
```diff
@@ -4238,16 +4238,18 @@ Namespace Microsoft.CodeAnalysis.VisualBasic
...
                     For Each expressionSyntax In node.Initializers
 
-                        Dim boundExpression As BoundExpression
-
-                        If expressionSyntax.Kind <> SyntaxKind.CollectionInitializer Then
-                            boundExpression = BindValue(expressionSyntax, diagnostics)
-
-                            If elemType IsNot Nothing Then
-                                boundExpression = ApplyImplicitConversion(expressionSyntax, elemType, boundExpression, diagnostics)
+                        Dim boundExpression As BoundExpression = BindValue(expressionSyntax, diagnostics)
```
Comment (333fred) [inline src/Compilers/VisualBasic/Portable/Binding/Binder_Expressions.vb]: Should use the temp DiagnosticsBag here too, previously the expressionSyntax wasn't being bound at all.

**Comment by 333fred** [comment]
Done review pass (commit 1)

File: src/Compilers/VisualBasic/Portable/Binding/Binder_Expressions.vb (hunk 1 above)
Comment (333fred) [reply src/Compilers/VisualBasic/Portable/Binding/Binder_Expressions.vb]: IMO there's no real value in reporting these nested diagnostics as they're not relevant to the issue the user is facing (too many dimensions). Technically this isn't a user-code breaking change as their code couldn't compile in the first place, but I don't see any value in reporting nested diagnostics in this scenario.

Comment (mavasani) [reply src/Compilers/VisualBasic/Portable/Binding/Binder_Expressions.vb]: So, I found couple of other related issues.

1. Few lines below this change, we have [another case](http://source.roslyn.io/#Microsoft.CodeAnalysis.VisualBasic/Binding/Binder_Expressions.vb,4270) where we report a diagnostic but don't bind the initializer if it is of unexpected syntax kind.
2. My current implementation leads to `BoundArrayLiteral` wrapped within a `BoundBadExpression` getting generated in the bound tree, but unlike other `BoundArrayLiteral` nodes which gets removed with [ReclassifyArrayLiteralExpression](http://source.roslyn.io/#Microsoft.CodeAnalysis.VisualBasic/Binding/Binder_Expressions.vb,1468), this one doesn't and this causes incorrect information from semantic model, such as the unit test failure [here](https://ci.dot.net/job/dotnet_roslyn/job/master/job/windows_debug_spanish_unit32_prtest/2240/testReport/junit/Microsoft.CodeAnalysis.VisualBasic.UnitTests.Semantics/ArrayLiteralTests/TestSemanticInfoTooDeeplyNestedArrayLiteral/).

I think we need a more involved binder fix here to handle all these cases properly, and that would be risky at this point. I am going to close my PR and move the bug to 16.0.

**Pull request closed without merging**